### PR TITLE
PT-323: PoC  the single abstract polymorph  json converter

### PR DESCRIPTION
**Pull request by tatarincev**
Jira: https://virtocommerce.atlassian.net/browse/PT-312

### CODE REVIEW

File: src/VirtoCommerce.Platform.Core/JsonConverters/PolymorphJsonConverter.cs [hunk 1]
```diff
@@ -1,49 +1,76 @@
...
-            object result;
-            var obj = JObject.Load(reader);
-            if (typeof(PermissionScope).IsAssignableFrom(objectType))
+
+            RegisterType(type, obj =>
             {
-                var scopeType = objectType.Name;
-                var pt = obj["type"] ?? obj["Type"];
+                var typeName = type.Name;
+                var pt = obj[descriminator] ?? obj[descriminator.FirstCharToUpper()];
```
Comment (eMazeika) [inline src/VirtoCommerce.Platform.Core/JsonConverters/PolymorphJsonConverter.cs:25]: consider using
var pt = obj.GetValue(discriminator, StringComparison.InvariantCultureIgnoreCase);

File: src/VirtoCommerce.Platform.Core/JsonConverters/PolymorphJsonConverter.cs [hunk 2]
```diff
@@ -1,49 +1,76 @@
...
         public override bool CanWrite => false;
         public override bool CanRead => true;
 
-        public override bool CanConvert(Type objectType)
-        {
-            return _knowTypes.Any(x => x.IsAssignableFrom(objectType));
-        }
 
-        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        public static void RegisterTypeForDescriminator(Type type, string descriminator)
```
Comment (eMazeika) [inline src/VirtoCommerce.Platform.Core/JsonConverters/PolymorphJsonConverter.cs:19]: fix typo in "discriminator"

File: src/VirtoCommerce.Platform.Core/JsonConverters/PolymorphJsonConverter.cs [hunk 3]
```diff
@@ -1,49 +1,76 @@
...
+            object result;
+            var obj = JObject.Load(reader);
+
+            object Factory(JObject obj2)
+            {
+                //TODO: Optmimize reflection
+                var tryCreateInstance = typeof(AbstractTypeFactory<>).MakeGenericType(objectType).GetMethods().FirstOrDefault(x => x.Name.Equals("TryCreateInstance") && x.GetParameters().Length == 0);
+                return tryCreateInstance?.Invoke(null, null);
+            };
+            var factory = _convertFactories.GetOrAdd(objectType, _ => Factory);
```
Comment (eMazeika) [inline src/VirtoCommerce.Platform.Core/JsonConverters/PolymorphJsonConverter.cs:70]: no need for named function, put the implementation as parameter. Same as for "_canConvertCache.GetOrAdd(objectType, _ =>"

**Review by eMazeika — CHANGES_REQUESTED** [review]
1. can you merge to dev branch
2. fix the 4 code smells

**Comment by Vectorfield4** [comment]
Please, change the target branch.

File: src/VirtoCommerce.Platform.Core/Domain/AbstractTypeFactory.cs [hunk 4]
```diff
@@ -23,6 +23,13 @@ public static IEnumerable<TypeInfo<BaseType>> AllTypeInfos
...
         }
 
+        public static bool HasOverrides
+        {
+            get
+            {
+                return _typeInfos.Count > 0;
+            }
+        }
         /// <summary>
```
Comment (ksavosteev) [inline src/VirtoCommerce.Platform.Core/Domain/AbstractTypeFactory.cs:33]: add empty line

File: src/VirtoCommerce.Platform.Core/JsonConverters/PolymorphJsonConverter.cs [hunk 5]
```diff
@@ -1,49 +1,76 @@
...
-                var pt = obj["type"] ?? obj["Type"];
+                var typeName = type.Name;
+                var pt = obj[descriminator] ?? obj[descriminator.FirstCharToUpper()];
                 if (pt != null)
                 {
-                    scopeType = pt.Value<string>();
+                    typeName = pt.Value<string>();
                 }
-                result = AbstractTypeFactory<PermissionScope>.TryCreateInstance(scopeType);
+                var tryCreateInstance = typeof(AbstractTypeFactory<>).MakeGenericType(type).GetMethods().FirstOrDefault(x => x.Name.Equals("TryCreateInstance") && x.GetParameters().Length == 1);
```
Comment (ksavosteev) [inline src/VirtoCommerce.Platform.Core/JsonConverters/PolymorphJsonConverter.cs:30]: can be extracted into method and resued (line 67)

File: src/VirtoCommerce.Platform.Core/JsonConverters/PolymorphJsonConverter.cs [hunk 6]
```diff
@@ -1,49 +1,108 @@
...
-using VirtoCommerce.Platform.Core.Security;
-using VirtoCommerce.Platform.Core.Settings;
 
 namespace VirtoCommerce.Platform.Core.JsonConverters
 {
     public class PolymorphJsonConverter : JsonConverter
     {
-        private static readonly Type[] _knowTypes = { typeof(ObjectSettingEntry), typeof(DynamicProperty), typeof(ApplicationUser), typeof(Role), typeof(PermissionScope) };
+        
+        public class CreateInstanceCacheKey
```
Comment (tatarincev) [inline src/VirtoCommerce.Platform.Core/JsonConverters/PolymorphJsonConverter.cs:13]: Why this code won't work for you? 
```
CacheKey.With(GetType(), type.Name, false);
```
I think this type `CreateInstanceCacheKey` is redundant for this task.

Comment (akak1977) [reply src/VirtoCommerce.Platform.Core/JsonConverters/PolymorphJsonConverter.cs]: Aggree, removed

File: src/VirtoCommerce.Platform.Web/Startup.cs [hunk 7]
```diff
@@ -518,6 +517,12 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
 
             // Use app insights telemetry 
             app.UseAppInsightsTelemetry();
+
+            
+            var mvcJsonOptions = app.ApplicationServices.GetService<IOptions<MvcNewtonsoftJsonOptions>>();
+            mvcJsonOptions.Value.SerializerSettings.Converters.Add(new PolymorphJsonConverter());
+            PolymorphJsonConverter.RegisterTypeForDiscriminator(typeof(PermissionScope), nameof(PermissionScope.Type));
+            PolymorphJsonConverter.RegisterTypeForDiscriminator(typeof(SearchCriteriaBase), nameof(SearchCriteriaBase.ObjectType));
```
Comment (eMazeika) [inline src/VirtoCommerce.Platform.Web/Startup.cs:525]: Remove

File: src/VirtoCommerce.Platform.Core/JsonConverters/PolymorphJsonConverter.cs [hunk 8]
```diff
@@ -1,49 +1,89 @@
...
+                var tryCreateInstance = _createInstanceMethodsCache.GetOrAdd(CacheKey.With(nameof(PolymorphJsonConverter), objectType.Name), _ =>
+                    typeof(AbstractTypeFactory<>).MakeGenericType(objectType).GetMethod("TryCreateInstance", 0 /* This guarantees template-parameterless method */, new Type[] { }));
+                return tryCreateInstance?.Invoke(null, null);
+            }); // Create instances for overrides and discriminator-less cases
+
+            result = factory(obj);
+
             serializer.Populate(obj.CreateReader(), result);
             return result;
         }
```
Comment (eMazeika) [inline src/VirtoCommerce.Platform.Core/JsonConverters/PolymorphJsonConverter.cs:89]: Move these 2 overridden methods up, right after CanRead, please

**Review by eMazeika — APPROVED** [review]

**Review by Vectorfield4 — APPROVED** [review]

File: src/VirtoCommerce.Platform.Web/Startup.cs (hunk 7 above)
Comment (akak1977) [reply src/VirtoCommerce.Platform.Web/Startup.cs:525]: SearchCriteriaBase is a platform class

Comment (tatarincev) [reply src/VirtoCommerce.Platform.Web/Startup.cs:525]: because it wasn't registered before

Comment (Vectorfield4) [reply src/VirtoCommerce.Platform.Web/Startup.cs:525]: https://github.com/VirtoCommerce/vc-platform/pull/2191

**Pull request merged**